### ISSUE TITLE
Define awaitCancel helper function to produce callback-induced streams

### ISSUE THREAD

**Issue by elizarov** [Kotlin/kotlinx.coroutines#289]
The implementation is trivial:

```
suspend fun awaitCancel(): Nothing = suspendCancellableCoroutine {}
```

The indented usage is to convert callback-based multi-fire APIs to channels like this:

```
fun produceXxx(): Channel<Xxx> = 
    // here context is an appropriate context where registerXxxCallback is allowed
    produce(context, capacity = Channel.CONFLATED) { // or Channel.UNLIMITED, -- depends on use-case
        try {
            registerXxxCallback { offer(it) } 
            awaitCancel()
        } finally { 
            unregisterXxxCallback()
        }
    }
```

**Comment by fvasco** [comment]
Why not use the `job` property (#260)?

```kotlin
fun produceXxx(): ReceiveChannel<Xxx> =
        Channel<Xxx>(Channel.UNLIMITED).apply {
            registerXxxCallback { offer(it) }
            job.invokeOnCompletion { unregisterXxxCallback() }
        }
```

**Comment by elizarov** [comment]
@fvasco Thanks. `job` property can subsume the need for `awaitCancel` in most (all?) cases. With the `job` property you should be also able to use it from inside of `produce`:

```
fun produceXxx(): Channel<Xxx> = 
    produce(capacity = Channel.CONFLATED) { 
        registerXxxCallback { offer(it) } 
        job.invokeOnCompletion { unregisterXxxCallback() }
    }
```

**Comment by fvasco** [comment]
I correct myself, the _right_ way is:

```kotlin
fun produceXxx(): ReceiveChannel<Xxx> =
        Channel<Xxx>(Channel.UNLIMITED).apply {
            val callback = XxxCallback { offer(it) }
            registerXxxCallback(callback)
            job.invokeOnCompletion { unregisterXxxCallback(callback) }
        }
```

**Comment by jcornaz** [comment]
Is it planned to add this? Or shall we prefer to use `invokeOnCompletion` of channel instead (like int @fvasco's example?

Personally I prefer the code style using `awaitCancel` i find it more natural, and less error-prone when callback registering/deregistering has to be done in specific thread (often the case with UI framework).

Here's how to get a click subscription Swing using `awaitCancel`:

```kotlin
fun Component.openMouseClickedSubscription(): ReceiveChannel<MouseEvent> =
    GlobalScope.produce(Dispatchers.Swing, Channel.UNLIMITED) {
      val listener = object : MouseAdapter() {
        override fun mouseClicked(e: MouseEvent) {
          offer(e)
        }
      }

      addMouseListener(listener) // called in EDT

      try {
        awaitCancel()
      } finally {
        removeMouseListener(listener) // called in EDT
      }
    }
```

And the same example using `invokeOnClose` instead of `awaitCancel` :
```kotlin
fun Component.openMouseClickedSubscription(): ReceiveChannel<MouseEvent> {
  val channel = Channel<MouseEvent>(Channel.UNLIMITED)

  val listener = object : MouseAdapter() {
    override fun mouseClicked(e: MouseEvent) {
      channel.offer(e)
    }
  }

  // this is needed if we want `openClickSubscription` to be callable from any thread
  SwingUtilities.invokeLater {

    addMouseListener(listener) // Has to be called in EDT

    channel.invokeOnClose {       // We cannot make any assumption about in which thread this will be executed

      SwingUtilities.invokeLater {
        removeMouseListener(listener) // Has to be called in EDT
      }
    }
  }

  return channel
}
```

Of course it is matter of personal taste. But I prefer the first one (using `awaitCancel`).

**Comment by Dico200** [comment]
`awaitCancellation` might be a slightly better name. 

The only point of invoking `awaitCancel` is when you want to keep coroutine alive until cancellation, and then run some code.

When the coroutines resumes, an exception will be thrown. So you need to wrap in try/finally block (which is better than `invokeOnCompletion` because the coroutine is able to resume in the same thread)

Maybe the two should be combined into one function.

**Comment by streetsofboston** [comment]
I suggest we'd name it `suspendForever` :-)

I like @Dico200 suggestion to combine the the two into one function:

`suspend fun suspendForever(onCompletion: (Throwable?) -> Unit): Nothing`

**Comment by LouisCAD** [comment]
I'd make onCompletion nullable in such a function, with a default value of
null.

**Comment by qwwdfsad** [comment]
`awaitClose` was implemented since `1.3.0`.
The actual pattern is even more elegant:
```
produce(context, capacity = Channel.CONFLATED) { // or Channel.UNLIMITED
    registerXxxCallback { offer(it) } 
     awaitClose { unregisterXxxCallback() }
}
```
